# Module 13: Pythonic code

There are four sections in Module 13.  The first three sections introduce list comprehension, f-strings, and lambda functions, respectively.  Comfort with these topics is one sign of an experienced Python programmer, and code that makes use of these sorts of Python-specific techniques is said to be *Pythonic*.

The fourth section in this notebook is a fun (and more advanced) example of using lambda functions and f-strings to change the way NumPy formats arrays.

## List comprehension

List comprehension has already been mentioned a few times, it represents a more "elegant" way of creating lists.  Actually, once you're comfortable with list comprehension, it feels very strange to make a list using a for-loop if list comprehension would also work.

In this section, we will learn this method more systematically.  When seeing an example of list comprehension, try to make the same list using for loops, and then compare the two methods, which will help you to obtain a better understanding.

**General structure** of a list comprehension:

```
[A for B in C]
```
where
* A is what you want to go in the list (object);
* B is a variable name (knob);
* C is the object you are iterating over (range).

### Example 1: Make a length 8 list of all 6s using list comprehension.

In this particular case,
* A is the number 6.
* B is `i`. (Any variable name would work and in this case it never gets used.  Underscore `_` would be another common choice.)
* C is `range(8)`.  In this case, the only important thing is that it has length 8.

In [99]:
[6 for i in range(8)]

[6, 6, 6, 6, 6, 6, 6, 6]

### Example 2: Let `mylist = [3,1,-2,10,-5,3,6,2,8]`.  Square each element in `mylist`.

In [100]:
mylist = [3,1,-2,10,-5,3,6,2,8]

In this case, the element that should go in the list is `x**2`, where `x` runs through the elements in `mylist`.

In [101]:
[x**2 for x in mylist]

[9, 1, 4, 100, 25, 9, 36, 4, 64]

What if we do not use every element in `mylist`? 

### Example 3: Get the sublist of `mylist` containing only the even numbers.

In [102]:
[x for x in mylist if x%2 == 0]

[-2, 10, 6, 2, 8]

In exmple 3, we used `if` conditional statement to select the elements. What if we need to consider two situations? Surely we can use an `if`-`else` conditional statement.

### Example 4: Replace each negative number in `mylist` with 0.

In [103]:
# this will only select the negative elements and replace them by 0
[0 for x in mylist if x < 0 ]

[0, 0]

So in this example, we need to use an `if`-`else` conditional statement, situation 1: an element is negative --> replace it by 0; situation 2: an element is non-negative --> keep it.

In [104]:
[x for x in mylist if x >= 0 else 0]

SyntaxError: invalid syntax (<ipython-input-104-1b54893fd78b>, line 1)

**If using the `if`-`else` conditional statement, we need to place it in front of the for loop!**

In [115]:
# "if x >= 0 else 0": if x is non-negative, x=x; else, x=0
[x if x >= 0 else 0 for x in mylist]

[3, 1, 0, 10, 0, 3, 6, 2, 8]

Here is another way to phrase it.

In [116]:
# "if x < 0 else x": if x is negative, x=0; else, x=x
[0 if x < 0 else x for x in mylist]

[3, 1, 0, 10, 0, 3, 6, 2, 8]

### Example 5: Make the length-5 list of lists `[[0,1,2], [0,1,2], ..., [0,1,2]]`.

It is very similar to generate a constant list.  In this case, the constant portion is the list `[0,1,2]`.

In [117]:
[[0,1,2] for _ in range(5)]

[[0, 1, 2], [0, 1, 2], [0, 1, 2], [0, 1, 2], [0, 1, 2]]

Each `[0,1,2]` is an element, so the length of this list is 8.

In [118]:
len([[0,1,2] for _ in range(5)])

5

### Example 6: Make the length-15 list `[0,1,2,0,1,2,...,0,1,2]`.

This is a little bit complicated, let's start from the for loop.

In [119]:
newlist = []
for j in range(5):
    for i in range(3):
        newlist.append(i)

In [120]:
newlist

[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]

This is the list comprehension version. 

`i` is the object we want to add into the list, and we have two for loops to control it.

In [121]:
[i for j in range(8) for i in range(3)]

[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]

We can also use list comprehension to deal with strings.

Here is an example where we work with a list of strings instead of a list of numbers.

### Example 7: Capitalize each word in the catalogue description of Math 9.

> Our Midterm 2 will be on December 6th (Wednesday), and will also be a 50-minute, in-class, written midterm. Do not miss it since there will be no make-up exam! (Sample midterm will be provided.)

If your string has line breaks, you should use triple apostrophes to surround it, like in the following.

In [122]:
s = '''Our Midterm 2 will be on December 6th, and will also be a 50-minute, in-class, written midterm. Do not miss 
it since there will be no make-up exam!'''

In [123]:
type(s)

str

If we try to directly iterate over the string, we will be iterating over the letters.

In [124]:
[c for c in s]

['O',
 'u',
 'r',
 ' ',
 'M',
 'i',
 'd',
 't',
 'e',
 'r',
 'm',
 ' ',
 '2',
 ' ',
 'w',
 'i',
 'l',
 'l',
 ' ',
 'b',
 'e',
 ' ',
 'o',
 'n',
 ' ',
 'D',
 'e',
 'c',
 'e',
 'm',
 'b',
 'e',
 'r',
 ' ',
 '6',
 't',
 'h',
 ',',
 ' ',
 'a',
 'n',
 'd',
 ' ',
 'w',
 'i',
 'l',
 'l',
 ' ',
 'a',
 'l',
 's',
 'o',
 ' ',
 'b',
 'e',
 ' ',
 'a',
 ' ',
 '5',
 '0',
 '-',
 'm',
 'i',
 'n',
 'u',
 't',
 'e',
 ',',
 ' ',
 'i',
 'n',
 '-',
 'c',
 'l',
 'a',
 's',
 's',
 ',',
 ' ',
 'w',
 'r',
 'i',
 't',
 't',
 'e',
 'n',
 ' ',
 'm',
 'i',
 'd',
 't',
 'e',
 'r',
 'm',
 '.',
 ' ',
 'D',
 'o',
 ' ',
 'n',
 'o',
 't',
 ' ',
 'm',
 'i',
 's',
 's',
 ' ',
 '\n',
 'i',
 't',
 ' ',
 's',
 'i',
 'n',
 'c',
 'e',
 ' ',
 't',
 'h',
 'e',
 'r',
 'e',
 ' ',
 'w',
 'i',
 'l',
 'l',
 ' ',
 'b',
 'e',
 ' ',
 'n',
 'o',
 ' ',
 'm',
 'a',
 'k',
 'e',
 '-',
 'u',
 'p',
 ' ',
 'e',
 'x',
 'a',
 'm',
 '!']

Here is an easy way to turn a string into a list of words.  (It separates the string at all white-space.)

In [125]:
wordlist = s.split()

In [126]:
wordlist[1]

'Midterm'

We eventually want to capitalize every word.  Strings in Python have a `capitalize` method.

In [127]:
wordlist[1].capitalize()

'Midterm'

The following will make a list that has each word capitalized.

In [128]:
caplist = [x.capitalize() for x in wordlist]

We used the following method once before to concatenate a list of strings together into one long string.  That's not quite what we want to do in this case, because we want spaces between the words.

In [129]:
''.join(caplist)

'OurMidterm2WillBeOnDecember6th,AndWillAlsoBeA50-minute,In-class,WrittenMidterm.DoNotMissItSinceThereWillBeNoMake-upExam!'

It's an easy change to get what we want.  Instead of starting with an empty string (length 0), we start with a length-1 string that contains a single space.  Then Python puts that space between all the words in our list.

In [130]:
' '.join(caplist)

'Our Midterm 2 Will Be On December 6th, And Will Also Be A 50-minute, In-class, Written Midterm. Do Not Miss It Since There Will Be No Make-up Exam!'

## f-strings

We often want to combine a fixed string with **the value of some variable**.  The preferred way to do this is to use f-strings, which are a relatively recent addition to Python.  Because f-strings are relatively new, you will often see older methods of doing the same thing, so it's important to be aware of the alternatives.

In [131]:
name = "all"
n = 10

Of course the following will not substitute the values for `name` or `n`.  (How would Python even know which `n`s to substitute?)

In [132]:
print("Dear all, I will curve down your grades by n points")

Dear all, I will curve down your grades by n points


**Method 1**: use `print` function.

In [133]:
print("Dear", name, ", I will curve down your grades by", n, "points")

Dear all , I will curve down your grades by 10 points


In very simple cases, using `+` to concatenate two strings is often the simplest method.  

But make sure the things you're combining really are **strings**. Otherwise you will get the following error.

In [134]:
print("Dear" + name + ", I will curve down your grades by" + n + "points")

TypeError: can only concatenate str (not "int") to str

The `+` version does work if we make sure to convert `n` to a string, using `str(n)`.

In [135]:
print("Dear" + name + ", I will curve down your grades by" + str(n) + "points")

Dearall, I will curve down your grades by10points


In [136]:
print("Dear" + name + ", I will curve down your grades by " + str(n) + " points")

Dearall, I will curve down your grades by 10 points


**Method 2**: use `format` function.

In more advanced cases, you will often see the `format` method being used; it might be the preferred way before f-strings showed up.  Empty curly brackets `{}` are used where a variable value should be substituted.  

One limitation of the `format` approach is that you need to keep track of the order of the variables that you want to include.

In [137]:
"Dear {}, I will curve down your grades by {} points".format(name, n)

'Dear all, I will curve down your grades by 10 points'

**Method 3**: the f-string method.

Here is finally the f-string approach.  Again curly brackets `{}` are used to indicate the portions Python should treat as an expression rather than as a string.  The f-string approach is much more readable than the earlier approaches.  Notice how we put an `f` before the quotation marks.

In [138]:
f"Dear {name}, I will curve down your grades by {n} points"

'Dear all, I will curve down your grades by 10 points'

There are lots of formatting options that can be used with f-strings and you can find them here: 

https://docs.python.org/3.4/library/string.html#format-specification-mini-language. 

In [139]:
m = 3**20

In [140]:
m

3486784401

In [141]:
f"{m}"

'3486784401'

If we would rather use the exponential notation (as we have seen in some NumPy outputs), we can use `:e`.

In [142]:
f"{m:e}"

'3.486784e+09'

If you want to see 4 places after the decimal point, you can use `:.4f`.  The `f` is saying to treat this number as a floating point number.

In [143]:
f"{m:.4f}"

'3486784401.0000'

In [144]:
f"{1/m:.4f}"

'0.0000'

Here we specify 15 decimal places.

In [145]:
f"{1/m:.15f}"

'0.000000000286797'

Notice that by default, Python displays `1/m` in exponential notation.

In [146]:
1/m

2.8679719907924413e-10

Here are some more formatting examples.  The following `:b` says to display the integer in **binary**.  For example, $17 = 1 \cdot 2^4 + 1 \cdot 2^0$, which is $10001$ when written in binary (also called "base 2").

In [147]:
z = 17

In [148]:
f"{z:b}"

'10001'

Here is the above number $3^{20}$ written in binary.

In [149]:
f"{m:b}"

'11001111110101000001101110010001'

## lambda functions

In this section we will introduce lambda functions, which are a concise way to define functions in Python.  They are very similar to function handles or anonymous functions in MATLAB.

### Example 8: Write a function `cap` that takes as input a string `s` and as output returns the same string `s` capitalized.

Our "usual" way of defining a function like `cap` uses the following syntax, but this syntax is an overkill for such a simple function.

In [150]:
def cap(s):
    return s.capitalize()

In [151]:
cap("hello there")

'Hello there'

Here is the lambda function approach: `lambda variable: return`   

We tell Python that we are about to define a function by using `lambda`.  The part that comes between `lambda` and the colon `:` specifies the variables for our function.  The part that comes after the colon is what the function will return.  The whole piece `lambda s: s.capitalize()` is defining the function, and we are saving that function with the name `cap`.

In [152]:
cap = lambda s: s.capitalize()

In [153]:
cap("hello there")

'Hello there'

### Example 9: Write a function `plus` that takes two inputs and adds them together.

In [154]:
plus = lambda x,y: x+y

In [155]:
plus(3,5)

8

The inputs can also be strings since they support `+`.

In [156]:
"hi" + "there"

'hithere'

In [157]:
plus("hi", "there")

'hithere'

### Example 10: Make a 20-by-3 NumPy array of random letters, then concatenate each row of three letters into a single length-3 string using `np.apply_along_axis`.

In [1]:
# preparation
import numpy as np
rng = np.random.default_rng()

Just as an aside, here is a way to get the 26 lower-case and 26 upper-case letters in English.

In [159]:
# yes, still preparation
import string

If we just want to see what is defined by this module `string`, we can use the `dir` function.  In the case of this module `string`, there aren't too many.

In [160]:
dir(string)

['Formatter',
 'Template',
 '_ChainMap',
 '_TemplateMetaclass',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_re',
 '_sentinel_dict',
 '_string',
 'ascii_letters',
 'ascii_lowercase',
 'ascii_uppercase',
 'capwords',
 'digits',
 'hexdigits',
 'octdigits',
 'printable',
 'punctuation',
 'whitespace']

On the other hand, if we check `dir(np)`, we will see many more options.

In [161]:
dir(np)

['ALLOW_THREADS',
 'AxisError',
 'BUFSIZE',
 'CLIP',
 'ComplexWarning',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'MachAr',
 'ModuleDeprecationWarning',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'Tester',
 'TooHardError',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'VisibleDeprecationWarning',
 'WRAP',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__dir__',
 '__doc__',
 '__file__',
 '__getattr__',
 '__git_revision__',
 '__loader__',
 '__mkl_version__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_ad

By scanning through `dir(string)`, you could imagine guessing and checking for the attribute we want.

In [162]:
string.ascii_letters

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

Now we know that `string.ascii_letters` returns to us all the letters. We can use `rng.choice` to randomly select letters to get random letters.

In [163]:
rng.choice(string.ascii_letters, size=(20,3))

ValueError: a must an array or an integer

The above error means the first input cannot be 'str'. So we convert `string.ascii_letters` into a list.

In [164]:
arr = rng.choice(list(string.ascii_letters), size=(20,3))
arr

array([['d', 'A', 'k'],
       ['X', 'D', 'J'],
       ['U', 'q', 'Y'],
       ['A', 'i', 'w'],
       ['G', 'r', 'X'],
       ['f', 'B', 'd'],
       ['N', 'u', 'q'],
       ['a', 'Y', 'C'],
       ['g', 's', 'F'],
       ['a', 'C', 'P'],
       ['m', 't', 'W'],
       ['h', 'L', 'N'],
       ['D', 'V', 'L'],
       ['i', 'd', 'q'],
       ['R', 'Y', 'A'],
       ['N', 'u', 'd'],
       ['D', 'D', 'o'],
       ['z', 'J', 'K'],
       ['S', 'Z', 'U'],
       ['X', 'e', 'X']], dtype='<U1')

Our goal is to take each of these 3-letter rows, and concatenate the letters together.  For example, we want to take `arr[3]` and get the string `"UvJ"`.

In [165]:
row = arr[3]

In [166]:
row

array(['A', 'i', 'w'], dtype='<U1')

Here is an example of what we want to do for the specific value `row`.

In [167]:
''.join(row)

'Aiw'

That is the function we want to apply to each row, so we will make a short version of it using a lambda function and pass that lambda function as an argument to `np.apply_along_axis`.  

The first three inputs of the function `np.apply_along_axis` are: (1) a function, (2) specify the axis/direction, and (3) specifies the array/source.

In [168]:
np.apply_along_axis(lambda row: ''.join(row), axis=1, arr=arr)

array(['dAk', 'XDJ', 'UqY', 'Aiw', 'GrX', 'fBd', 'Nuq', 'aYC', 'gsF',
       'aCP', 'mtW', 'hLN', 'DVL', 'idq', 'RYA', 'Nud', 'DDo', 'zJK',
       'SZU', 'XeX'], dtype='<U3')

If we were going to be using this function `lambda row: ''.join(row)` repeatedly, it would be better to name the function, like in the following.

In [169]:
join_letters = lambda row: ''.join(row)

If we had `join_letters` defined that way, then we could replace our code above with `np.apply_along_axis(lambda row: ''.join(row), axis=1, arr=arr)`.

In [170]:
np.apply_along_axis(join_letters, axis=1, arr=arr)

array(['dAk', 'XDJ', 'UqY', 'Aiw', 'GrX', 'fBd', 'Nuq', 'aYC', 'gsF',
       'aCP', 'mtW', 'hLN', 'DVL', 'idq', 'RYA', 'Nud', 'DDo', 'zJK',
       'SZU', 'XeX'], dtype='<U3')

### Example 11: Let `tuplist` be the following list of tuples.  Sort this list so that the numbers are increasing.

```[("A", 40), ("B", 60), ("E", 30), ("C", 20), ("D", 45)]```

In [171]:
tuplist = [("A", 40), ("B", 60), ("E", 30), ("C", 20), ("D", 45)]

We are going to sort this length-5 list using Python's `sorted` function.  Notice that `sorted` accepts a `key` keyword argument.  The argument passed as `key` should be a function, and then that function will be applied before the sorting is done.

In [172]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In our case, we want to sort by the numbers.

In [173]:
tup = ("B", 60)

In [174]:
tup[1]

60

The following is the right idea, but we need to use the `key =` portion of the keyword argument.  (When there are multiple options that we can sort, we need to use thekeyword arguments to specify which one we want to select.)

In [175]:
sorted(tuplist, lambda tup: tup[1])

TypeError: sorted expected 1 argument, got 2

Notice how this sorts the tuples `tup` according to the number `tup[1]`.

In [176]:
sorted(tuplist, key = lambda tup: tup[1])

[('C', 20), ('E', 30), ('A', 40), ('D', 45), ('B', 60)]

There is one other keyword argument that we can see in the documentation above.  If we want the values to be decreasing instead of increasing, we should specify `reverse = True`.

In [177]:
sorted(tuplist, key = lambda tup: tup[1], reverse = True)

[('B', 60), ('D', 45), ('A', 40), ('E', 30), ('C', 20)]

As one last example, Python is equally happy sorting by letters rather than numbers.  Here we sort using `key = lambda tup: tup[0]`.

In [178]:
sorted(tuplist, key = lambda tup: tup[0])

[('A', 40), ('B', 60), ('C', 20), ('D', 45), ('E', 30)]

## Using f-strings and a lambda function to format a NumPy array

The goal in this section is to apply some of the earlier concepts to a more advanced (and more specialized) example.  We show how to change the way NumPy formats the numbers it displays.  That particular application is not the important part.  The reason we're using this application is because it involves f-strings and lambda functions (as well as NumPy arrays, dictionaries, data types, ...).

In [179]:
import numpy as np

Notice that when we create our random number generator in the next cell, we specify a `seed` argument.  So the numbers that are produced are reproducible (if you run it multiple times, you will get the same result, which is the seed).

In [180]:
rng = np.random.default_rng(seed=1)
arr = rng.normal(size=100)
arr

array([ 0.3456,  0.8216,  0.3304, -1.3032,  0.9054,  0.4464, -0.5370,
        0.5811,  0.3646,  0.2941,  0.0284,  0.5467, -0.7365, -0.1629,
       -0.4821,  0.5988,  0.0397, -0.2925, -0.7819, -0.2572,  0.0081,
       -0.2756,  1.2941,  1.0067, -2.7112, -1.8890, -0.1748, -0.4222,
        0.2136,  0.2173,  2.1178, -1.1120, -0.3776,  2.0428,  0.6467,
        0.6631, -0.5140, -1.6481,  0.1675,  0.1090, -1.2274, -0.6832,
       -0.0720, -0.9448, -0.0983,  0.0955,  0.0356, -0.5063,  0.5937,
        0.8912,  0.3208, -0.8182,  0.7317, -0.5014,  0.8792, -1.0718,
        0.9145, -0.0201, -1.2487, -0.3139,  0.0541,  0.2728, -0.9822,
       -1.1074,  0.1996, -0.4667,  0.2355,  0.7595, -1.6488,  0.2544,
        1.2246, -0.2975, -0.8108,  0.7522,  0.2534,  0.8959, -0.3452,
       -1.4818, -0.1100, -0.4458,  0.7753,  0.1936, -1.6308, -1.1952,
        0.8838,  0.6798, -0.6402, -0.0010,  0.4456,  0.4684,  0.8762,
        0.2565, -0.0948, -0.2588,  1.0557, -2.2509, -0.1387,  0.0330,
       -1.4253,  0.3

Notice how the following numbers are exactly the same as before when we also used `seed=1`.

One option (probably not the easiest option) to prevent NumPy from using scientific notation is to use the NumPy's `set_printoptions` function. This accepts many different arguments (notice how much longer this documentation is than the documentation for `sorted` above). The argument we are going to use is the `formatter` argument.

The `formatter` argument is supposed to be a 
> dict of callables ... the keys should indicate the type(s)...  Callables should return a string...

In particular, the argument is supposed to be a dictionary.  The keys should be data types (or rather, strings representing the data types), and the values in the dictionary are supposed to be "callables", which we can think of as functions, that return a string.

In [181]:
help(np.set_printoptions)

Help on function set_printoptions in module numpy:

set_printoptions(precision=None, threshold=None, edgeitems=None, linewidth=None, suppress=None, nanstr=None, infstr=None, formatter=None, sign=None, floatmode=None, *, legacy=None)
    Set printing options.
    
    These options determine the way floating point numbers, arrays and
    other NumPy objects are displayed.
    
    Parameters
    ----------
    precision : int or None, optional
        Number of digits of precision for floating point output (default 8).
        May be None if `floatmode` is not `fixed`, to print as many digits as
        necessary to uniquely specify the value.
    threshold : int, optional
        Total number of array elements which trigger summarization
        rather than full repr (default 1000).
        To always use the full repr without summarization, pass `sys.maxsize`.
    edgeitems : int, optional
        Number of array items in summary at beginning and end of
        each dimension (default 

Here is a specific example.  The dictionary we pass has just one key in it, `'float'`.  The value is a function, and we write that function using a lambda function.  The input to the function will be a float, and the output of the function is supposed to be a string; in this case, the string should be how we want floats to be displayed.

Here is a first attempt specifying `formatter`.  We will use string formatting options to specify that we want 4 decimal places.

In [182]:
np.set_printoptions(formatter= {'float': (lambda x: f"{x:.4f}")})

We are still specifying `seed=1`.  Notice how the initial value in the array, `3.45584192e-01` above, now gets displayed as a decimal number with four digits, `0.3456`.

In [183]:
rng = np.random.default_rng(seed=1)
arr = rng.normal(size=100)
arr

array([0.3456, 0.8216, 0.3304, -1.3032, 0.9054, 0.4464, -0.5370, 0.5811,
       0.3646, 0.2941, 0.0284, 0.5467, -0.7365, -0.1629, -0.4821, 0.5988,
       0.0397, -0.2925, -0.7819, -0.2572, 0.0081, -0.2756, 1.2941, 1.0067,
       -2.7112, -1.8890, -0.1748, -0.4222, 0.2136, 0.2173, 2.1178,
       -1.1120, -0.3776, 2.0428, 0.6467, 0.6631, -0.5140, -1.6481, 0.1675,
       0.1090, -1.2274, -0.6832, -0.0720, -0.9448, -0.0983, 0.0955,
       0.0356, -0.5063, 0.5937, 0.8912, 0.3208, -0.8182, 0.7317, -0.5014,
       0.8792, -1.0718, 0.9145, -0.0201, -1.2487, -0.3139, 0.0541, 0.2728,
       -0.9822, -1.1074, 0.1996, -0.4667, 0.2355, 0.7595, -1.6488, 0.2544,
       1.2246, -0.2975, -0.8108, 0.7522, 0.2534, 0.8959, -0.3452, -1.4818,
       -0.1100, -0.4458, 0.7753, 0.1936, -1.6308, -1.1952, 0.8838, 0.6798,
       -0.6402, -0.0010, 0.4456, 0.4684, 0.8762, 0.2565, -0.0948, -0.2588,
       1.0557, -2.2509, -0.1387, 0.0330, -1.4253, 0.3328])

Let's see some other ways to change this float formatting, so that the numbers get more nicely lined up.  

In this particular case, we indicate that there should be a `+` sign used when the numbers are positive.

In [184]:
np.set_printoptions(formatter= {'float': (lambda x: f"{x:+.4f}")})

The numbers are lined up nicely in the following, but the `+` signs still look a little strange.

In [185]:
rng = np.random.default_rng(seed=1)
arr = rng.normal(size=100)
arr

array([+0.3456, +0.8216, +0.3304, -1.3032, +0.9054, +0.4464, -0.5370,
       +0.5811, +0.3646, +0.2941, +0.0284, +0.5467, -0.7365, -0.1629,
       -0.4821, +0.5988, +0.0397, -0.2925, -0.7819, -0.2572, +0.0081,
       -0.2756, +1.2941, +1.0067, -2.7112, -1.8890, -0.1748, -0.4222,
       +0.2136, +0.2173, +2.1178, -1.1120, -0.3776, +2.0428, +0.6467,
       +0.6631, -0.5140, -1.6481, +0.1675, +0.1090, -1.2274, -0.6832,
       -0.0720, -0.9448, -0.0983, +0.0955, +0.0356, -0.5063, +0.5937,
       +0.8912, +0.3208, -0.8182, +0.7317, -0.5014, +0.8792, -1.0718,
       +0.9145, -0.0201, -1.2487, -0.3139, +0.0541, +0.2728, -0.9822,
       -1.1074, +0.1996, -0.4667, +0.2355, +0.7595, -1.6488, +0.2544,
       +1.2246, -0.2975, -0.8108, +0.7522, +0.2534, +0.8959, -0.3452,
       -1.4818, -0.1100, -0.4458, +0.7753, +0.1936, -1.6308, -1.1952,
       +0.8838, +0.6798, -0.6402, -0.0010, +0.4456, +0.4684, +0.8762,
       +0.2565, -0.0948, -0.2588, +1.0557, -2.2509, -0.1387, +0.0330,
       -1.4253, +0.3

Another string formatting option is to leave a blank space when the number is positive.  I find this to be the version which is easiest to read.

In [186]:
np.set_printoptions(formatter= {'float': (lambda x: f"{x: .4f}")})

In [187]:
rng = np.random.default_rng(seed=1)
arr = rng.normal(size=100)
arr

array([ 0.3456,  0.8216,  0.3304, -1.3032,  0.9054,  0.4464, -0.5370,
        0.5811,  0.3646,  0.2941,  0.0284,  0.5467, -0.7365, -0.1629,
       -0.4821,  0.5988,  0.0397, -0.2925, -0.7819, -0.2572,  0.0081,
       -0.2756,  1.2941,  1.0067, -2.7112, -1.8890, -0.1748, -0.4222,
        0.2136,  0.2173,  2.1178, -1.1120, -0.3776,  2.0428,  0.6467,
        0.6631, -0.5140, -1.6481,  0.1675,  0.1090, -1.2274, -0.6832,
       -0.0720, -0.9448, -0.0983,  0.0955,  0.0356, -0.5063,  0.5937,
        0.8912,  0.3208, -0.8182,  0.7317, -0.5014,  0.8792, -1.0718,
        0.9145, -0.0201, -1.2487, -0.3139,  0.0541,  0.2728, -0.9822,
       -1.1074,  0.1996, -0.4667,  0.2355,  0.7595, -1.6488,  0.2544,
        1.2246, -0.2975, -0.8108,  0.7522,  0.2534,  0.8959, -0.3452,
       -1.4818, -0.1100, -0.4458,  0.7753,  0.1936, -1.6308, -1.1952,
        0.8838,  0.6798, -0.6402, -0.0010,  0.4456,  0.4684,  0.8762,
        0.2565, -0.0948, -0.2588,  1.0557, -2.2509, -0.1387,  0.0330,
       -1.4253,  0.3

There is definitely much more that we could learn involving "Pythonic" code; this has just been a quick introduction to some of the most important ways to make your code more Pythonic.  The three most important concepts introduced in this section were list comprehension, f-strings, and lambda functions.